### DISCUSSION TITLE
How to avoid page updates?

### DISCUSSION THREAD

**Discussion by me21** [zauberzeug/nicegui#41] [Q&A] [answered]
Do I understand correctly that any event handler (upload, click, timer, etc.) by default leads to full page update and the handler should return False to prevent that?

**Comment by falkoschindler** [comment]
Yes, that's a rather non-obvious "feature" NiceGUI adopts from JustPy: To avoid a complete page update after handling an event, the event handler should return `False` (or anything but `None`).

Right now I can't find JustPy's documentation about that. But here is a corresponding comment in the source code:
https://github.com/elimintz/justpy/blob/42275e9c6e00373f09affd7d781a278383171807/justpy/justpy.py#L356

```python
    # If page is not to be updated, the event_function should return anything but None
    if event_result is None:
        if com_type == 0:     # WebSockets communication
            if LATENCY:
                await asyncio.sleep(LATENCY / 1000)
            await p.update()
        elif com_type == 1:   # Ajax communication
            build_list = p.build_list()
```

Usually the default behavior should be fine, but for frequent events (or timer callbacks) it's reasonable to return `False` where possible.

**Reply by me21** [reply]
I wonder in which case would we need full page update. If new components should be added to the page?

**Reply by falkoschindler** [reply]
I'm not sure. Right now we're thinking about changing the default behavior to "no page update". We could allow for a return value `True` to trigger an update anyway if it is explicitly required. But the default should be more intuitive.

**Comment by falkoschindler** [comment, marked as answer]
Since NiceGUI 0.8.0 this is no longer the case. Event handlers don't automatically update the whole page. Only elements that are directly modified (e.g. their text, value, style, classes or props) are sent to the client. In edge cases you might need to trigger an explicit `ui.update` for individual elements.
